lib/SIL/IR/SILInstruction.cpp

```diff
@@ -1295,17 +1295,57 @@ bool SILInstruction::isDeallocatingStack() const {
 bool SILInstruction::mayRequirePackMetadata() const {
   switch (getKind()) {
   case SILInstructionKind::AllocPackInst:
-  case SILInstructionKind::PartialApplyInst:
-  case SILInstructionKind::ApplyInst:
-  case SILInstructionKind::BeginApplyInst:
-  case SILInstructionKind::TryApplyInst:
-  case SILInstructionKind::DebugValueInst:
-  case SILInstructionKind::MetatypeInst:
   case SILInstructionKind::TuplePackElementAddrInst:
   case SILInstructionKind::OpenPackElementInst:
-  case SILInstructionKind::ClassMethodInst:
-  case SILInstructionKind::WitnessMethodInst:
     return true;
+  case SILInstructionKind::PartialApplyInst:
+  case SILInstructionKind::ApplyInst:
+  case SILInstructionKind::BeginApplyInst:
+  case SILInstructionKind::TryApplyInst: {
+    // Check the function type for packs.
+    auto apply = ApplySite::isa(const_cast<SILInstruction *>(this));
+    if (apply.getCallee()->getType().isOrContainsPack(*getFunction()))
+      return true;
+    // Check the function signature for packs.
```

Comment:
Doesn't this need to look at the substitution map? You might have something like this

```
apply %f<G<repeat each T>>(...)
```